### DISCUSSION TITLE
How to run the examples in a docker container ?

### DISCUSSION THREAD

**Discussion by EricBoix** [iTowns/itowns#1631] [Q&A] [answered]
## The Environment
* Version used: v2.26.0
* Browser Name and version: curl
* Operating System and version (desktop or mobile): OS ubuntu:focal, npm version 6.14.4
* Link to your project: https://github.com/VCityTeam/UD-Viz/

## Context
I wish to run my iTowns based application within a docker container. As starting point
I first try to run any iTowns example within an Ubuntu "focal" docker container. 
Getting (some) iTowns container to build and run is quite out the box. Yet when
http querrying the container no answer seems to be returned. 

Warning: this difficulty could be independent from iTowns and only related to
misunderstood docker port mapping ways of things (my docker mileage is low). 

## Steps to Reproduce
```
# Create a docker context
mkdir docker-itowns
pushd docker-itowns

# Write a Dockerfile
echo "FROM ubuntu:focal"                        >  Dockerfile
echo "RUN apt-get update"                       >> Dockerfile
echo "RUN apt-get install -y nodejs"        >> Dockerfile
echo "RUN apt-get install -y npm"            >> Dockerfile
echo "RUN apt-get install -y git"               >> Dockerfile
echo "RUN git clone https://github.com/iTowns/itowns.git  --branch  v2.26.0"        >> Dockerfile
echo "WORKDIR /itowns"                         >> Dockerfile
echo "RUN npm install"                            >> Dockerfile
echo "EXPOSE 8080"                              >> Dockerfile
echo 'CMD [ "npm", "start" ]'                  >> Dockerfile
popd

# Build the image:
docker build -t itowns docker-itowns

# Launch the build container in detached mode:
docker run -d -p 8080:8080/tcp -p 8080:8080/udp  -t itowns:latest
    <your_container_id>    #  ----> a returned container id

# checking that the container is properly running: should return 
# somethings of the form
#  "itowns:latest  "npm start"  <...>   0.0.0.0:8080->8080/tcp, 0.0.0.0:8080->8080/udp
docker ps                         

# Check (again) that port mapping is well done
docker port <your_container_id>
```

The above tends to show that the container is running and listening of 
the host localhost 8080 port properly mapped to the containers internal 8080
port.

## Expected Behavior
 Sending an http request to the dockerized iTowns e.g.
```
curl -i localhost:8080
```
should return some http answer.

## Actual Behavior
Curl gets no answer that is the command
```
curl -i localhost:8080
```
returns "curl: (52) Empty reply from server"

Note: the same request executed inside the container runs properly i.e.
with the above detached running container
```
docker exec -it <your_container_id>  /bin/bash
   root:/itowns# apt-get install curl
   root:/itowns# curl -i localhost:8080
```
does return an http answer

## Possible Cause/Fix/Solution
Possible causes:
 - A misunderstanding of docker port mapping as illustrated by [this StackOverFlow thread](https://stackoverflow.com/questions/26677502/docker-expose-cant-get-it) or described [in this tutorial](https://www.bitcoininsider.org/article/32223/dockerfile-expose-how-define-ports-docker-containers) ?
 - A misunderstanding of what containerizing a node.js application requires (refer e.g. to [this tutorial](https://nodejs.org/fr/docs/guides/nodejs-docker-webapp/)) ?

**Comment by gchoqueux** [comment]
have you solved your problem
I don't know enough about Docker

I'm going to inquire @arnaudbirk

**Comment by gchoqueux** [comment, marked as answer]
We have solution, thanks @arnaudbirk

It's not docker problem, could you try:

```
# Write a Dockerfile
FROM node:lts-stretch
RUN apt-get -y update
RUN apt-get install -y git
RUN git clone https://github.com/iTowns/itowns.git  --branch  v2.26.0
WORKDIR /itowns
RUN npm install
RUN npm run prepare
EXPOSE 8080
CMD [ "npm", "start", "--", "--host=0.0.0.0" ]

# build
docker build --no-cache -t docker-itowns 

#start 
docker run -ti --rm -p 8080:8080 docker-itowns
```

**Comment by EricBoix** [comment]
Yep, thanks @arnaudbirk and @gchoqueux : that does the trick for me !

### Note
I took me quite some time to be able to successfully reproduce this solution
because, at first, I dropped the the port redirection (from 8080 to 8080).
That is I ran
```
docker run -ti --rm  docker-itowns
```
and this did NOT work for me. 

Hence port redirection (repetition or exposure might be better is this case?) is a must
for the trick to work.

### Changing the port
This side note might be obvious to knowledgeable docker users.

If you additionally wish to use another port than the default 8080, if suffice to
adapt the EXPOSE and CMD within the above Dockerfile, that is to finish it
with e.g.
```
# Dockerfile
...                    
RUN npm run prepare      # As above

EXPOSE 8999
CMD [ "npm", "start", "--", "--host=0.0.0.0", "--port=8999" ]
```
and, after rebuilding the container, run it with
```
docker run -t --rm -p 8999:8999 docker-itowns
``` 

As far as I'm concerned this issue can be closed.
Thanks again.
